{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a935ac11-b8e5-4114-8e20-315625ffe8ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PyTorch 2.5.1+cu124\n",
      "CUDA available: True (NVIDIA GeForce RTX 4070 Ti SUPER)\n",
      "Pandas 2.2.3\n",
      "tensorqtl 1.0.10\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "import pandas as pd\n",
    "import torch\n",
    "import tensorqtl\n",
    "from tensorqtl import pgen, cis\n",
    "import time\n",
    "import os\n",
    "import os.path as op\n",
    "import sys\n",
    "import glob\n",
    "import re\n",
    "print(f'PyTorch {torch.__version__}')\n",
    "#print(torch.__version__)\n",
    "print('CUDA available: {} ({})'.format(torch.cuda.is_available(), torch.cuda.get_device_name(torch.cuda.current_device())))\n",
    "print(f'Pandas {pd.__version__}')\n",
    "print(f'tensorqtl {tensorqtl.__version__}')\n",
    "\n",
    "outdir='eqtl_out'\n",
    "if not os.path.exists(outdir):\n",
    "    os.makedirs(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b769386-4358-46be-ba6f-7bcffa0c4a98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Plink2 pgen file found.\n",
      "Features found:  ['gene']\n",
      " Change the `features` array below to set the features to be processed:\n",
      "Features to process:  ['gene']\n"
     ]
    }
   ],
   "source": [
    "## ---- CHANGE HERE --------------\n",
    "in_dir='eqtl_inputs' ## directory path where the tensorqtl input files can be found:\n",
    "                     ## dsname.gene.expr.bed.gz and dsname.covars.txt\n",
    "dsname='habenula' ## dataset name, this is the prefix for input and output file names\n",
    "plink='genotypes/habenula_maf05' ## plink prefix for genotype data\n",
    "#plink='genotypes/mdd_maf01'\n",
    "mapping_file=None ## comment the line below if genotype IDs are the same with RNAseq sample IDs.\n",
    "#mapping_file='genoID2rnaID.tab' # genotype IDs will be changed to their RNAseq mappings in the 2nd column\n",
    "#mapping_file='genotypes/mdd_geno2rna.tab'\n",
    "## if not None, mapping_file must be the same with the one used for 02_prep_tensorQTL_RSE_input.R\n",
    "col_map = None\n",
    "if mapping_file is not None and op.exists(mapping_file):\n",
    "    print(f\"mapping genotype IDs to RNAseq sample IDs based on mapping file: {mapping_file}\")\n",
    "    mapping_df = pd.read_csv(mapping_file, sep=\"\\t\", header=None, names=[\"gt_id\", \"r_id\"])\n",
    "    col_map = dict(zip(mapping_df['gt_id'], mapping_df['r_id']))\n",
    "if op.exists(plink+'.pgen'): # checks for plink2 pgen file for genotype data\n",
    "    print(\"Plink2 pgen file found.\")\n",
    "else:\n",
    "    raise Exception(\"Plink2 \"+plink+\".pgen not found!\")\n",
    "\n",
    "exprfiles=glob.glob(in_dir+'/'+dsname+'.*.expr.bed.gz')\n",
    "features = [os.path.basename(file).split('.')[1] for file in exprfiles]\n",
    "exprdict = {os.path.basename(file).split('.')[1]: file for file in exprfiles}\n",
    "for feat in features:\n",
    "    expres = exprdict[feat]\n",
    "    covar=expres.replace(\"expr.bed.gz\", \"covars.txt\")\n",
    "    if not op.exists(covar):\n",
    "       raise FileNotFoundError(error_message)(\"Covars file \"+covar+\" not found!\")\n",
    "print(\"Features found: \",features)\n",
    "print(\" Change the `features` array below to set the features to be processed:\")\n",
    "## CHANGE here and uncomment, if needed, in order to select the features to process:\n",
    "features = ['gene']\n",
    "print(\"Features to process: \", features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7c26c66a-f599-447e-8250-5caa4b743dd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotype dimensions: (5504021, 69)\n"
     ]
    }
   ],
   "source": [
    "#pr = genotypeio.PlinkReader(plink)\n",
    "## use plink2 pgen format\n",
    "pgr = pgen.PgenReader(plink)\n",
    "genotype_df = pgr.load_genotypes()\n",
    "#variant_df = pr.bim.set_index('snp')[['chrom', 'pos']]\n",
    "variant_df = pgr.variant_df\n",
    "print(\"Genotype dimensions: \", end='')\n",
    "print(genotype_df.shape)\n",
    "# if a mapping file is given\n",
    "# Check if all genotype_df column names are in the mapping file\n",
    "#all_columns_found = genotype_df.columns.isin(mapping_df[\"current_column\"]).all()\n",
    "\n",
    "if col_map is not None:\n",
    "  # Rename columns using the dictionary\n",
    "  genotype_df.rename(columns=col_map, inplace=True)\n",
    "## Now genotype_df contains the updated column names\n",
    "#print(genotype_df.iloc[:5, :7])  # Display the first few rows\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d66fd02a-988f-4a2d-8565-373ab505a2f4",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " Processing feature: gene\n",
      "Covariates dim:(68, 19)\n",
      "Phenotype dimensions:(22756, 68)\n",
      "These genotypes have no expression data given: {'Br5572'}\n",
      "cis-QTL mapping: nominal associations for all variant-phenotype pairs\n",
      "  * 68 samples\n",
      "  * 22756 phenotypes\n",
      "  * 19 covariates\n",
      "  * 5504021 variants\n",
      "  * applying in-sample 0.05 MAF filter\n",
      "  * cis-window: ±500,000\n",
      "    ** dropping 65 phenotypes on chrs. without genotypes\n",
      "  * checking phenotypes: 22691/22691\n",
      "    ** dropping 140 phenotypes without variants in cis-window\n",
      "  * Computing associations\n",
      "    Mapping chromosome chr1\n",
      "    time elapsed: 0.04 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr2\n",
      "    time elapsed: 0.09 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr3\n",
      "    time elapsed: 0.14 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr4\n",
      "    time elapsed: 0.17 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr5\n",
      "    time elapsed: 0.20 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr6\n",
      "    time elapsed: 0.24 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr7\n",
      "    time elapsed: 0.28 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr8\n",
      "    time elapsed: 0.31 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr9\n",
      "    time elapsed: 0.34 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr10\n",
      "    time elapsed: 0.37 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr11\n",
      "    time elapsed: 0.41 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr12\n",
      "    time elapsed: 0.45 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr13\n",
      "    time elapsed: 0.47 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr14\n",
      "    time elapsed: 0.49 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr15\n",
      "    time elapsed: 0.52 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr16\n",
      "    time elapsed: 0.55 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr17\n",
      "    time elapsed: 0.59 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr18\n",
      "    time elapsed: 0.61 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr19\n",
      "    time elapsed: 0.65 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr20\n",
      "    time elapsed: 0.68 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr21\n",
      "    time elapsed: 0.69 min\n",
      "    * writing output\n",
      "    Mapping chromosome chr22\n",
      "    time elapsed: 0.70 min\n",
      "    * writing output\n",
      "    Mapping chromosome chrX\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/tensorqtl/lib/python3.11/site-packages/tensorqtl/core.py:153: UserWarning: var(): degrees of freedom is <= 0. Correction should be strictly less than the reduction factor (input numel divided by output numel). (Triggered internally at ../aten/src/ATen/native/ReduceOps.cpp:1823.)\n",
      "  genotype_var_t = genotype_res_t.var(1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    time elapsed: 0.72 min\n",
      "    * writing output\n",
      "done.\n",
      "All done.\n"
     ]
    }
   ],
   "source": [
    "def fixBrNums(col):\n",
    "    return re.sub(r'^Br(\\d\\d\\d)$', r'Br0\\1', col)\n",
    "\n",
    "## fix Brnums with 3 digits\n",
    "genotype_df.columns = [fixBrNums(col) for col in genotype_df.columns]\n",
    "\n",
    "\n",
    "##  etc. Fix chromosomes (add the \"chr\" prefix) if needed:\n",
    "if not variant_df.chrom.iloc[0].startswith('chr'):\n",
    "   variant_df.chrom = [ 'chr' + chrom for chrom in variant_df.chrom]\n",
    "## select chromosomes - to make sure we have the same chromosomes in our data for each expression dataset\n",
    "variant_chrom = set(variant_df.chrom)\n",
    "\n",
    "for feat in features:\n",
    "    print(f\" Processing feature: {feat}\")\n",
    "    tag = dsname+'.'+feat\n",
    "    expres = exprdict[feat]\n",
    "    covar=expres.replace(\"expr.bed.gz\", \"covars.txt\")\n",
    "    covariates_df = pd.read_csv(covar, sep='\\t', index_col=0).T\n",
    "    ## fix Brnums with 3 digits\n",
    "    covariates_df.index = [fixBrNums(idx) for idx in covariates_df.index]\n",
    "\n",
    "    print(\"Covariates dim:\", end='')\n",
    "    print(covariates_df.shape)\n",
    "    #print(covariates_df.iloc[:5, :5])\n",
    "    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expres)\n",
    "    print(\"Phenotype dimensions:\", end='')\n",
    "    print(phenotype_df.shape)\n",
    "    ## fix Brnums with 3 digits\n",
    "    phenotype_df.columns = [fixBrNums(col) for col in phenotype_df.columns]\n",
    "\n",
    "\n",
    "    ## use the same chromosome set\n",
    "    express_chrom = set(phenotype_pos_df.chr)\n",
    "    assert len(variant_chrom.intersection(express_chrom))>0\n",
    "    if express_chrom - variant_chrom:\n",
    "        chrom_filter = phenotype_pos_df.chr.isin(variant_chrom)\n",
    "        if (len(chrom_filter)<phenotype_df.shape[0]):\n",
    "          phenotype_df = phenotype_df[chrom_filter]\n",
    "          phenotype_pos_df = phenotype_pos_df[chrom_filter]\n",
    "    ## make sure we keep only the genotypes for the same expression samples\n",
    "    cols=phenotype_df.columns.tolist()\n",
    "    gcols=genotype_df.columns.tolist()\n",
    "    ## which genotypes are in the expression data? print them here if any are missing and stop the process\n",
    "    missing_geno = set(cols) - set(gcols)\n",
    "    if missing_geno:\n",
    "        print(f\"Genotypes missing for {len(missing_geno)} samples: {missing_geno}\")\n",
    "        raise RuntimeError(\"Missing genotypes detected. Aborting...\")\n",
    "    ## show if any genotypes are not in the expression data:\n",
    "    missing_expr = set(gcols) - set(cols)\n",
    "    if missing_expr:\n",
    "        print(f\"These genotypes have no expression data given: {missing_expr}\")\n",
    "    geno_df=genotype_df.loc[:, cols]\n",
    "    ## run tensorQTL:\n",
    "    cis.map_nominal(geno_df, variant_df, phenotype_df, phenotype_pos_df, prefix = tag, covariates_df= covariates_df,\n",
    "                maf_threshold=0.05, window=500000, output_dir= outdir, verbose=False)\n",
    "print(\"All done.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2bd5e7cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Phenotype columns: ['Br6323', 'Br1016', 'Br2421', 'Br1637', 'Br1842', 'Br0983', 'Br1203', 'Br1738', 'Br1507', 'Br1682', 'Br1383', 'Br1427', 'Br1416', 'Br1425', 'Br1487', 'Br1526', 'Br2292', 'Br1676', 'Br2378', 'Br2589', 'Br6104', 'Br6070', 'Br8048', 'Br6158', 'Br5873', 'Br5234', 'Br5891', 'Br5488', 'Br5581', 'Br5319', 'Br5398', 'Br5412', 'Br5573', 'Br5446', 'Br5888', 'Br1034', 'Br1092', 'Br1204', 'Br1378', 'Br1469', 'Br1565', 'Br1735', 'Br1761', 'Br2015', 'Br2080', 'Br2476', 'Br1023', 'Br1980', 'Br2044', 'Br2052', 'Br2425', 'Br5292', 'Br5385', 'Br5459', 'Br5555', 'Br5558', 'Br5639', 'Br5702', 'Br5712', 'Br5756', 'Br5871', 'Br6197', 'Br6264', 'Br8050', 'Br8218', 'Br5212', 'Br1350', 'Br1225']\n",
      "Covariates index: ['Br6323', 'Br1016', 'Br2421', 'Br1637', 'Br1842', 'Br983', 'Br1203', 'Br1738', 'Br1507', 'Br1682', 'Br1383', 'Br1427', 'Br1416', 'Br1425', 'Br1487', 'Br1526', 'Br2292', 'Br1676', 'Br2378', 'Br2589', 'Br6104', 'Br6070', 'Br8048', 'Br6158', 'Br5873', 'Br5234', 'Br5891', 'Br5488', 'Br5581', 'Br5319', 'Br5398', 'Br5412', 'Br5573', 'Br5446', 'Br5888', 'Br1034', 'Br1092', 'Br1204', 'Br1378', 'Br1469', 'Br1565', 'Br1735', 'Br1761', 'Br2015', 'Br2080', 'Br2476', 'Br1023', 'Br1980', 'Br2044', 'Br2052', 'Br2425', 'Br5292', 'Br5385', 'Br5459', 'Br5555', 'Br5558', 'Br5639', 'Br5702', 'Br5712', 'Br5756', 'Br5871', 'Br6197', 'Br6264', 'Br8050', 'Br8218', 'Br5212', 'Br1350', 'Br1225']\n"
     ]
    }
   ],
   "source": [
    "# Print columns of phenotype_df and index of covariates_df for debugging\n",
    "print(\"Phenotype columns:\", phenotype_df.columns.tolist())\n",
    "print(\"Covariates index:\", covariates_df.index.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c0e9c393-844b-4ad7-8178-791d51d3d3c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Br6323</th>\n",
       "      <th>Br1016</th>\n",
       "      <th>Br2421</th>\n",
       "      <th>Br1637</th>\n",
       "      <th>Br1842</th>\n",
       "      <th>Br0983</th>\n",
       "      <th>Br1203</th>\n",
       "      <th>Br1738</th>\n",
       "      <th>Br1507</th>\n",
       "      <th>Br1682</th>\n",
       "      <th>...</th>\n",
       "      <th>Br5712</th>\n",
       "      <th>Br5756</th>\n",
       "      <th>Br5871</th>\n",
       "      <th>Br6197</th>\n",
       "      <th>Br6264</th>\n",
       "      <th>Br8050</th>\n",
       "      <th>Br8218</th>\n",
       "      <th>Br5212</th>\n",
       "      <th>Br1350</th>\n",
       "      <th>Br1225</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>-2.171784</td>\n",
       "      <td>-1.837840</td>\n",
       "      <td>-0.411353</td>\n",
       "      <td>-0.382676</td>\n",
       "      <td>-1.126355</td>\n",
       "      <td>-1.108615</td>\n",
       "      <td>-2.333769</td>\n",
       "      <td>-1.184872</td>\n",
       "      <td>-1.403475</td>\n",
       "      <td>-0.928317</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.792600</td>\n",
       "      <td>-1.625827</td>\n",
       "      <td>-1.697315</td>\n",
       "      <td>-1.014606</td>\n",
       "      <td>-0.021202</td>\n",
       "      <td>-1.782940</td>\n",
       "      <td>-0.509058</td>\n",
       "      <td>-1.119866</td>\n",
       "      <td>-1.295406</td>\n",
       "      <td>-1.880302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>1.062472</td>\n",
       "      <td>1.551763</td>\n",
       "      <td>1.736710</td>\n",
       "      <td>2.369074</td>\n",
       "      <td>1.384064</td>\n",
       "      <td>1.690661</td>\n",
       "      <td>1.704066</td>\n",
       "      <td>1.359794</td>\n",
       "      <td>1.963830</td>\n",
       "      <td>2.500164</td>\n",
       "      <td>...</td>\n",
       "      <td>1.828595</td>\n",
       "      <td>1.076911</td>\n",
       "      <td>2.074495</td>\n",
       "      <td>1.671447</td>\n",
       "      <td>2.219662</td>\n",
       "      <td>0.986524</td>\n",
       "      <td>1.593278</td>\n",
       "      <td>2.311974</td>\n",
       "      <td>1.520791</td>\n",
       "      <td>0.378487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000279457.3</th>\n",
       "      <td>2.471219</td>\n",
       "      <td>2.521775</td>\n",
       "      <td>2.088978</td>\n",
       "      <td>2.160759</td>\n",
       "      <td>2.300490</td>\n",
       "      <td>2.309100</td>\n",
       "      <td>1.647845</td>\n",
       "      <td>1.919035</td>\n",
       "      <td>2.725320</td>\n",
       "      <td>2.592680</td>\n",
       "      <td>...</td>\n",
       "      <td>1.860885</td>\n",
       "      <td>1.701429</td>\n",
       "      <td>2.345663</td>\n",
       "      <td>2.379377</td>\n",
       "      <td>2.561471</td>\n",
       "      <td>1.152476</td>\n",
       "      <td>2.785751</td>\n",
       "      <td>2.422692</td>\n",
       "      <td>1.988977</td>\n",
       "      <td>2.241173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000228463.9</th>\n",
       "      <td>1.199063</td>\n",
       "      <td>2.210713</td>\n",
       "      <td>1.474934</td>\n",
       "      <td>1.703086</td>\n",
       "      <td>0.965648</td>\n",
       "      <td>1.154444</td>\n",
       "      <td>1.433372</td>\n",
       "      <td>0.778913</td>\n",
       "      <td>2.120458</td>\n",
       "      <td>1.009240</td>\n",
       "      <td>...</td>\n",
       "      <td>0.465975</td>\n",
       "      <td>-0.831128</td>\n",
       "      <td>1.577355</td>\n",
       "      <td>1.755960</td>\n",
       "      <td>1.707017</td>\n",
       "      <td>2.125105</td>\n",
       "      <td>1.651893</td>\n",
       "      <td>2.008958</td>\n",
       "      <td>1.520791</td>\n",
       "      <td>-0.063116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000236679.2</th>\n",
       "      <td>-1.219832</td>\n",
       "      <td>-0.212262</td>\n",
       "      <td>-0.817122</td>\n",
       "      <td>-0.495877</td>\n",
       "      <td>0.069631</td>\n",
       "      <td>-0.026137</td>\n",
       "      <td>-0.274160</td>\n",
       "      <td>-0.998200</td>\n",
       "      <td>0.144013</td>\n",
       "      <td>-0.595708</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.967343</td>\n",
       "      <td>-0.321649</td>\n",
       "      <td>-0.634523</td>\n",
       "      <td>-1.325979</td>\n",
       "      <td>-0.255546</td>\n",
       "      <td>-0.910134</td>\n",
       "      <td>-1.917174</td>\n",
       "      <td>-0.498810</td>\n",
       "      <td>-0.965249</td>\n",
       "      <td>-1.484700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Br6323    Br1016    Br2421    Br1637    Br1842    Br0983  \\\n",
       "ID                                                                              \n",
       "ENSG00000278267.1 -2.171784 -1.837840 -0.411353 -0.382676 -1.126355 -1.108615   \n",
       "ENSG00000227232.5  1.062472  1.551763  1.736710  2.369074  1.384064  1.690661   \n",
       "ENSG00000279457.3  2.471219  2.521775  2.088978  2.160759  2.300490  2.309100   \n",
       "ENSG00000228463.9  1.199063  2.210713  1.474934  1.703086  0.965648  1.154444   \n",
       "ENSG00000236679.2 -1.219832 -0.212262 -0.817122 -0.495877  0.069631 -0.026137   \n",
       "\n",
       "                     Br1203    Br1738    Br1507    Br1682  ...    Br5712  \\\n",
       "ID                                                         ...             \n",
       "ENSG00000278267.1 -2.333769 -1.184872 -1.403475 -0.928317  ... -0.792600   \n",
       "ENSG00000227232.5  1.704066  1.359794  1.963830  2.500164  ...  1.828595   \n",
       "ENSG00000279457.3  1.647845  1.919035  2.725320  2.592680  ...  1.860885   \n",
       "ENSG00000228463.9  1.433372  0.778913  2.120458  1.009240  ...  0.465975   \n",
       "ENSG00000236679.2 -0.274160 -0.998200  0.144013 -0.595708  ... -1.967343   \n",
       "\n",
       "                     Br5756    Br5871    Br6197    Br6264    Br8050    Br8218  \\\n",
       "ID                                                                              \n",
       "ENSG00000278267.1 -1.625827 -1.697315 -1.014606 -0.021202 -1.782940 -0.509058   \n",
       "ENSG00000227232.5  1.076911  2.074495  1.671447  2.219662  0.986524  1.593278   \n",
       "ENSG00000279457.3  1.701429  2.345663  2.379377  2.561471  1.152476  2.785751   \n",
       "ENSG00000228463.9 -0.831128  1.577355  1.755960  1.707017  2.125105  1.651893   \n",
       "ENSG00000236679.2 -0.321649 -0.634523 -1.325979 -0.255546 -0.910134 -1.917174   \n",
       "\n",
       "                     Br5212    Br1350    Br1225  \n",
       "ID                                               \n",
       "ENSG00000278267.1 -1.119866 -1.295406 -1.880302  \n",
       "ENSG00000227232.5  2.311974  1.520791  0.378487  \n",
       "ENSG00000279457.3  2.422692  1.988977  2.241173  \n",
       "ENSG00000228463.9  2.008958  1.520791 -0.063116  \n",
       "ENSG00000236679.2 -0.498810 -0.965249 -1.484700  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phenotype_df.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tensorqtl",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
