---
title: "eqtl_gather_results"
##output: html_notebook
---


```{r}
library(arrow)
library(data.table)
library(BiocParallel)
library(qs)
library(SummarizedExperiment)
library(stringi)
#library(biomaRt) # for getting dbSNP info about specific SNPs
```

```{r}
############# CHANGE HERE ############
## path to parquest files, INCLUDING the dataset prefix (e.g. prefix.gene.cis_qtl_pairs.chr1.parquet)
dsname="amyg" ## same dataset name value used in previous steps
parquet_dir = 'mdd_old/tensorQTL_lognorm/eqtl_out' ## expected there: amyg.gene.cis_qtl_pairs.chr1.parquet
out_dir = 'eqtl_fdr05' # output directory to write FDR<0.05 eqtl mappings
##############################

pat='cis_qtl_pairs\\.chr1\\.parquet$'
dsfeat <- list.files(path=parquet_dir, pattern=pat, recursive = T, full.names = F)
dsfeat <- grep(paste0('^\\Q',dsname,'\\E'), dsfeat, value=T, perl=T)
dsfeat <- stri_replace_all_regex(dsfeat, pat, '')
#ds <- unique(stri_split_fixed(regfeat, '.', simplify=T)[,1])
feats <- unique(stri_split_fixed(dsfeat, '.', simplify=T)[,2])
if (length(feats)==0) stop("No features found in", parquet_dir, ' for dataset prefix: ', dsname)
```

```{r}
if (!dir.exists(out_dir)) dir.create(out_dir)
fedts <- list() # eqtl data tables per feature
for (feature in feats) {
     rfprefix=paste0(dsname, '.', feature)
     feqtls=paste0(out_dir, '/', rfprefix, '.genomewide_FDR05.qs')
     if (!file.exists(feqtls)) {
        pfiles <- list.files(path=parquet_dir, pattern=paste0(rfprefix,'\\.cis_qtl_pairs.chr\\d+\\.parquet$'),
                             recursive = T, full.names = T)
        cat("   ", length(pfiles), "parquet files found for", rfprefix, "\n")
        stopifnot(length(pfiles)<=23 & length(pfiles)>0)
        dlst <- bplapply(pfiles, function(f) {
           pdt <- read_parquet(f)
           setDT(pdt)
           pdt[, fdr:=p.adjust(pval_nominal, "fdr")]
           pdt <- na.omit(pdt, cols='pval_nominal')
           setorder(pdt, fdr)
           sigdt <- pdt[fdr<=0.05,]
           rm(pdt)
           return(sigdt)
         }, BPPARAM=MulticoreParam(8))
         fedts[[feature]] <- rbindlist(dlst)
         cat(" ..saving", feqtls, "\n")
         qsave(fedts[[feature]], file=feqtls)
        } else {
          cat(" ..loading", feqtls, "\n")
          fedts[[feature]] <- qread(feqtls)
        }
}
```


